File: backend/api/schemas/hypothesis.py
from pydantic import BaseModel
from typing import Dict, List, Optional

# The four tests the original Gradio app supports, verbatim.
ONE_SAMPLE_T = "One sample Student's t-test"
TWO_SAMPLE_T = "Two samples Student's t-test"
EQUAL_VARIANCE = "Equal variance between two groups"
ONE_WAY_ANOVA = "One-way ANOVA"


class GroupSpec(BaseModel):
    """A group is the union of one or more selected values of one categorical column.

    Group 1 and group 2 may name different columns — that flexibility is inherited
    from the Gradio app and is deliberate.
    """
    column: str
    values: List[str]
    name: str = "Group"


class HypothesisParams(BaseModel):
    session_id: str
    column: str
    test_type: str
    alpha: float = 0.05
    alternative: str = "two-sided"
    mu0: Optional[float] = None
    correction: bool = True              # Welch correction for the two-sample t-test
    variance_test_type: str = "Levene"   # "Bartlett" | "Levene"
    group1: Optional[GroupSpec] = None
    group2: Optional[GroupSpec] = None
    anova_column: Optional[str] = None
    anova_levels: Optional[List[str]] = None
    filters: Optional[Dict[str, List[str]]] = None


class RejectionRegionData(BaseModel):
    """Mirrors ``core.hypothesis_testing.rejection_region`` 1:1."""
    dist: str
    dof: List[float]
    x: List[float]
    pdf: List[float]
    y_max: float
    x_range: List[float]
    statistic: float
    statistic_offscale: bool
    critical_values: List[float]
    reject_region: List[List[float]]
    p_area: List[List[float]]
    tail: str
    reject: bool


class GroupSummary(BaseModel):
    name: str
    n: int
    mean: float
    sd: float
    var: float


class HypothesisResponse(BaseModel):
    test_type: str
    table: str  # JSON string of the raw result DataFrame via to_json(orient="records")
    statistic: float
    p_value: float
    alpha: float
    reject: bool
    verdict: str
    h0: str
    h1: str
    rejection_region: RejectionRegionData
    group_summary: Optional[List[GroupSummary]] = None
    warnings: List[str] = []

File: frontend/src/api/hypothesis.ts
const BASE_URL = "/api/hypothesis";

export const ONE_SAMPLE_T = "One sample Student's t-test";
export const TWO_SAMPLE_T = "Two samples Student's t-test";
export const EQUAL_VARIANCE = "Equal variance between two groups";
export const ONE_WAY_ANOVA = "One-way ANOVA";

export const TEST_TYPES = [
  ONE_SAMPLE_T,
  TWO_SAMPLE_T,
  EQUAL_VARIANCE,
  ONE_WAY_ANOVA,
] as const;

export type TestType = (typeof TEST_TYPES)[number];
export type Alternative = "two-sided" | "greater" | "less";

export interface GroupSpec {
  column: string;
  values: string[];
  name: string;
}

export interface HypothesisParams {
  session_id: string;
  column: string;
  test_type: string;
  alpha: number;
  alternative: Alternative;
  mu0?: number | null;
  correction: boolean;
  variance_test_type: string;
  group1?: GroupSpec | null;
  group2?: GroupSpec | null;
  anova_column?: string | null;
  anova_levels?: string[] | null;
  filters?: Record<string, string[]> | null;
}

export interface RejectionRegionData {
  dist: "t" | "chi2" | "f";
  dof: number[];
  x: number[];
  pdf: number[];
  /** Ceiling for the y axis: crops the spike a one-degree-of-freedom density has near zero. */
  y_max: number;
  /** Axis domain: sized to the null distribution, not to an extreme statistic. */
  x_range: [number, number];
  statistic: number;
  /** True when the statistic lies past `x_range` and its marker must be pinned to the edge. */
  statistic_offscale: boolean;
  critical_values: number[];
  /** Shaded because of α. */
  reject_region: number[][];
  /** Shaded because of the observed statistic. */
  p_area: number[][];
  tail: Alternative;
  reject: boolean;
}

export interface GroupSummary {
  name: string;
  n: number;
  mean: number;
  sd: number;
  var: number;
}

export interface HypothesisResponse {
  test_type: string;
  /** JSON string of the raw library result table; columns differ per test. */
  table: string;
  statistic: number;
  p_value: number;
  alpha: number;
  reject: boolean;
  verdict: string;
  h0: string;
  h1: string;
  rejection_region: RejectionRegionData;
  group_summary?: GroupSummary[] | null;
  warnings: string[];
}

export const hypothesisApi = {
  runTest: async (params: HypothesisParams): Promise<HypothesisResponse> => {
    const response = await fetch(`${BASE_URL}/test`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "x-session-id": params.session_id,
      },
      body: JSON.stringify(params),
    });
    if (!response.ok) {
      const text = await response.text();
      let errStr = text;
      try { errStr = JSON.parse(text).detail || text; } catch {}
      throw new Error(errStr);
    }
    return response.json();
  },
};
